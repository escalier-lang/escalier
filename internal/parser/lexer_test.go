package parser

import (
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func TestMain(m *testing.M) {
	v := m.Run()
	snaps.Clean(m) // remove unused snapshots
	os.Exit(v)
}

func TestLexingKeywords(t *testing.T) {
	source := Source{
		Path:     "input.esc",
		Contents: "fn var val",
	}

	lexer := NewLexer(source)

	snaps.MatchSnapshot(t, lexer.Lex())
}

func TestLexingOperators(t *testing.T) {
	source := Source{
		Path:     "input.esc",
		Contents: "+ - * / =",
	}

	lexer := NewLexer(source)

	snaps.MatchSnapshot(t, lexer.Lex())
}

func TestLexingIdentifiers(t *testing.T) {
	source := Source{
		Path:     "input.esc",
		Contents: "foo\nbar",
	}

	lexer := NewLexer(source)

	snaps.MatchSnapshot(t, lexer.Lex())
}

func TestLexingLiterals(t *testing.T) {
	source := Source{
		Path:     "input.esc",
		Contents: "\"hello\"",
	}

	lexer := NewLexer(source)

	snaps.MatchSnapshot(t, lexer.Lex())
}

func TestLexingParens(t *testing.T) {
	source := Source{
		Path:     "input.esc",
		Contents: "a * (b + c)",
	}

	lexer := NewLexer(source)

	snaps.MatchSnapshot(t, lexer.Lex())
}
