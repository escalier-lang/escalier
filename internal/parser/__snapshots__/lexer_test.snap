
[TestLexingKeywords - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:3},
            SourceID: 0,
        },
        Type:  14,
        Value: "fn",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:4},
            End:      ast.Location{Line:1, Column:7},
            SourceID: 0,
        },
        Type:  18,
        Value: "var",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:8},
            End:      ast.Location{Line:1, Column:11},
            SourceID: 0,
        },
        Type:  19,
        Value: "val",
    },
}
---

[TestLexingOperators - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:2},
            SourceID: 0,
        },
        Type:  39,
        Value: "+",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:3},
            End:      ast.Location{Line:1, Column:4},
            SourceID: 0,
        },
        Type:  40,
        Value: "-",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:5},
            End:      ast.Location{Line:1, Column:6},
            SourceID: 0,
        },
        Type:  41,
        Value: "*",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:7},
            End:      ast.Location{Line:1, Column:8},
            SourceID: 0,
        },
        Type:  42,
        Value: "/",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:9},
            End:      ast.Location{Line:1, Column:10},
            SourceID: 0,
        },
        Type:  47,
        Value: "=",
    },
}
---

[TestLexingIdentifiers - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:4},
            SourceID: 0,
        },
        Type:  0,
        Value: "foo",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:2, Column:1},
            End:      ast.Location{Line:2, Column:4},
            SourceID: 0,
        },
        Type:  0,
        Value: "bar",
    },
}
---

[TestLexingLiterals - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:8},
            SourceID: 0,
        },
        Type:  2,
        Value: "hello",
    },
}
---

[TestLexingParens - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:2},
            SourceID: 0,
        },
        Type:  0,
        Value: "a",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:3},
            End:      ast.Location{Line:1, Column:4},
            SourceID: 0,
        },
        Type:  41,
        Value: "*",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:5},
            End:      ast.Location{Line:1, Column:6},
            SourceID: 0,
        },
        Type:  66,
        Value: "(",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:6},
            End:      ast.Location{Line:1, Column:7},
            SourceID: 0,
        },
        Type:  0,
        Value: "b",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:8},
            End:      ast.Location{Line:1, Column:9},
            SourceID: 0,
        },
        Type:  39,
        Value: "+",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:10},
            End:      ast.Location{Line:1, Column:11},
            SourceID: 0,
        },
        Type:  0,
        Value: "c",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:11},
            End:      ast.Location{Line:1, Column:12},
            SourceID: 0,
        },
        Type:  67,
        Value: ")",
    },
}
---

[TestLexingLineComments - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:1, Column:7},
            SourceID: 0,
        },
        Type:  10,
        Value: "// foo",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:2, Column:1},
            End:      ast.Location{Line:2, Column:7},
            SourceID: 0,
        },
        Type:  10,
        Value: "// bar",
    },
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:3, Column:1},
            End:      ast.Location{Line:3, Column:1},
            SourceID: 0,
        },
        Type:  73,
        Value: "",
    },
}
---

[TestLexingBlockComment - 1]
[]*parser.Token{
    &parser.Token{
        Span: ast.Span{
            Start:    ast.Location{Line:1, Column:1},
            End:      ast.Location{Line:4, Column:2},
            SourceID: 0,
        },
        Type:  11,
        Value: "/**\n * foo\n * bar\n */",
    },
}
---
